### PR TITLE
Add options parameter that replaces strict_kwonly

### PR DESCRIPTION
**Pull request by mcarans**
The strict_kwonly flag only affects parameters that have a default value - when set to False, parameters with default values are accepted as keyword arguments instead of postional arguments.

Deprecate strict_kwonly and replace by options="has_default" (i.e. strict_kwonly=False), ="kwonly" (strict_kwonly=True), "all" (all parameters accepted as keyword arguments)

### CODE REVIEW

File: lib/defopt.py [hunk 1]
```diff
@@ -196,14 +202,21 @@ def run(funcs: Union[Callable, List[Callable], Dict[str, Callable]], *,
         dashes) to letters, to use as alternative short flags.  Defaults to
         `None`, which means to generate short flags for any non-ambiguous
         option.  Set to ``{}`` to completely disable short flags.
+    :param options:
+        If `all`, all parameters are converted into command-line flags. If
```
Comment (anntzer) [inline lib/defopt.py:206]: Use single quotes, not backquotes, to delimit strings (all, has_default, kwonly).  Two spaces after dots.

File: lib/defopt.py [hunk 2]
```diff
@@ -196,14 +202,21 @@ def run(funcs: Union[Callable, List[Callable], Dict[str, Callable]], *,
...
+        `has_default`, all parameters with a default are converted into
+        command-line flags. The default behavior (`kwonly`) is to convert
+        keyword-only parameters to command line flags, and non-keyword-only
+        parameters with a default to optional positional command line
+        parameters.
     :param strict_kwonly:
-        If `False`, all parameters with a default are converted into
-        command-line flags.  The default behavior (`True`) is to convert
+        Deprecated. If `False`, all parameters with a default are converted
+        into command-line flags. The default behavior (`True`) is to convert
```
Comment (anntzer) [inline lib/defopt.py:214]: ditto re: space.

File: lib/defopt.py [hunk 3]
```diff
@@ -368,14 +382,24 @@ def signature(func: Callable):
         parameters=parameters, return_annotation=return_annotation)
 
 
-def _populate_parser(func, parser, parsers, short, strict_kwonly,
+def _populate_parser(func, parser, parsers, short, options, strict_kwonly,
                      show_defaults, show_types, no_negated_flags):
+    if strict_kwonly is _unset:
```
Comment (anntzer) [inline lib/defopt.py:387]: I would put this check at the top of the definition of bind(), immediately normalizing to `options` and not bothering to pass `strict_kwonly` down the chain anywhere else.

File: lib/defopt.py [hunk 4]
```diff
@@ -196,14 +202,21 @@ def run(funcs: Union[Callable, List[Callable], Dict[str, Callable]], *,
         dashes) to letters, to use as alternative short flags.  Defaults to
         `None`, which means to generate short flags for any non-ambiguous
         option.  Set to ``{}`` to completely disable short flags.
+    :param options:
```
Comment (anntzer) [inline lib/defopt.py:205]: Upong further consideration, "options" is too vague a name, let's go for "cli_options" (unless you have a better proposal)

File: lib/defopt.py [hunk 5]
```diff
@@ -368,14 +382,24 @@ def signature(func: Callable):
...
+        warnings.warn(msg, DeprecationWarning)
+        options = 'kwonly' if strict_kwonly else 'has_default'
     sig = signature(func)
     doc = _parse_docstring(inspect.getdoc(func))
     parser.description = doc.text
 
     positionals = {name for name, param in sig.parameters.items()
-                   if ((param.default is param.empty or strict_kwonly)
+                   if (((param.default is param.empty and options == 'has_default')
+                        or options == 'kwonly')
```
Comment (anntzer) [inline lib/defopt.py:402]: I would list options == 'kwonly' first (as it's the default)

File: test_defopt.py [hunk 6]
```diff
@@ -307,6 +307,14 @@ def test_bool_kwarg(self):
...
             return foo
+        self.assertIs(defopt.run(main, options='has_default',
+                                 argv=[]), 'default')
+        self.assertIs(defopt.run(main, options='has_default',
+                                 argv=['--foo']), True)
+        self.assertIs(defopt.run(main, options='has_default',
+                                 argv=['--no-foo']), False)
+        self.assertIs(defopt.run(main, options='has_default',
+                                 argv=['--foo', '--no-foo']), False)
         self.assertIs(defopt.run(main, strict_kwonly=False,
```
Comment (anntzer) [inline test_defopt.py:318]: I would additionally wrap all of these with self.assertWarns(DeprecationWarning): ...

**Comment by anntzer** [comment]
As a side point, looks like a recent setuptools release broke CI.  Let's see whether they fix this quickly, otherwise I'll have to fiddle with the CI config :/

**Comment by anntzer** [comment]
I bumped the dependency to docutils>=0.12 to fix the build, but you'll need to rebase the PR.

**Comment by mcarans** [comment]
@anntzer Thanks for all the good suggestions. I've done them all. Please take a look.

File: lib/defopt.py [hunk 7]
```diff
@@ -196,14 +211,21 @@ def run(funcs: Union[Callable, List[Callable], Dict[str, Callable]], *,
         dashes) to letters, to use as alternative short flags.  Defaults to
         `None`, which means to generate short flags for any non-ambiguous
         option.  Set to ``{}`` to completely disable short flags.
+    :param cli_options:
+        If ``all``, all parameters are converted into command-line flags.  If
```
Comment (anntzer) [inline lib/defopt.py:215]: again, please use plain single quotes (`'`) in the docs to specify strings

File: lib/defopt.py [hunk 8]
```diff
@@ -170,7 +184,8 @@ def bind(funcs: Union[Callable, List[Callable], Dict[str, Callable]], *,
 def run(funcs: Union[Callable, List[Callable], Dict[str, Callable]], *,
         parsers: Dict[type, Callable[[str], Any]] = {},
         short: Optional[Dict[str, str]] = None,
-        strict_kwonly: bool = True,
+        cli_options: str = _unset,
```
Comment (anntzer) [inline lib/defopt.py:187]: I guess the typehint should be `Literal['kwonly', 'all', 'has_default']`; ditto for bind.  (I think I like this order)

File: lib/defopt.py [hunk 9]
```diff
@@ -196,14 +211,21 @@ def run(funcs: Union[Callable, List[Callable], Dict[str, Callable]], *,
         dashes) to letters, to use as alternative short flags.  Defaults to
         `None`, which means to generate short flags for any non-ambiguous
         option.  Set to ``{}`` to completely disable short flags.
+    :param cli_options:
+        If ``all``, all parameters are converted into command-line flags.  If
+        ``has_default``, all parameters with a default are converted into
+        command-line flags.  The default behavior (``kwonly``) is to convert
+        keyword-only parameters to command line flags, and non-keyword-only
+        parameters with a default to optional positional command line
+        parameters.
```
Comment (anntzer) [inline lib/defopt.py:220]: I think I prefer reordering to start with the default ("The default behavior ('kwonly') is ....  'all' turns all parameters into command-line flags; 'has_default' turns a parameter into a command-line flag if and only if it has a default value.")

**Comment by mcarans** [comment]
@anntzer I've made the chanegs you suggested. Please take a look.

File: lib/defopt.py [hunk 10]
```diff
@@ -170,7 +184,8 @@ def bind(funcs: Union[Callable, List[Callable], Dict[str, Callable]], *,
 def run(funcs: Union[Callable, List[Callable], Dict[str, Callable]], *,
         parsers: Dict[type, Callable[[str], Any]] = {},
         short: Optional[Dict[str, str]] = None,
-        strict_kwonly: bool = True,
+        cli_options:  Literal['kwonly', 'all', 'has_default'] = _unset,
```
Comment (anntzer) [inline lib/defopt.py:187]: double spaced

**Comment by anntzer** [comment]
Just one typo, then good to go  Do you want to take care of squashing your commits yourself or should I do it?

**Comment by mcarans** [comment]
@anntzer  I have fixed the double space. I'd be grateful if you could squash the commits.

**Comment by anntzer** [comment]
Merged into master as https://github.com/anntzer/defopt/commit/6507c5380becc717aa7c521aff5246d3561087a4.  Thanks for the PR!

**Pull request closed without merging**